File: src/converter/plugins/NamespaceInfoPlugin.ts
import {ConverterPlugin} from "../plugin.js";
import {ModuleDeclaration, Node} from "typescript";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {NamespaceStrategy} from "../../configuration/configuration.js";
import {NamespaceInfo} from "../../structure/namespace/collectNamespaceInfo.js";
import {extractNamespaceName} from "../../structure/namespace/createNamespaceInfoItem.js";

export const namespaceInfoServiceKey = Symbol()

export class NamespaceInfoService {
    private readonly namespaceInfo: Record<string, NamespaceStrategy> = {}


    constructor(namespaceInfo: NamespaceInfo) {
        namespaceInfo.forEach(item => {
            this.namespaceInfo[item.name] = item.strategy
        })
    }

    resolveNamespaceStrategy(node: ModuleDeclaration): NamespaceStrategy {
        const name = extractNamespaceName(node)
        const detailedName = name.map(it => it.detailedName).join(".")

        return this.namespaceInfo[detailedName]
    }
}

export class NamespaceInfoPlugin implements ConverterPlugin {
    private readonly namespaceInfoService: NamespaceInfoService;

    constructor(namespaceInfo: NamespaceInfo) {
        this.namespaceInfoService = new NamespaceInfoService(namespaceInfo);
    }

    generate(): Record<string, string> {
        return {};
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    traverse(node: Node): void {
    }

    setup(context: ConverterContext): void {
        context.registerService(namespaceInfoServiceKey, this.namespaceInfoService)
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/NamespaceInfoPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.configuration.NamespaceStrategy
import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.ConverterPlugin
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.structure.namespace.NamespaceInfo
import io.github.sgrishchenko.karakum.structure.namespace.extractNamespaceName
import js.symbol.Symbol
import typescript.ModuleDeclaration
import typescript.Node

@OptIn(ExperimentalJsExport::class)
@JsExport
val namespaceInfoServiceKey = Symbol()

@OptIn(ExperimentalJsExport::class)
@JsExport
class NamespaceInfoService @JsExport.Ignore constructor(namespaceInfo: NamespaceInfo) {
    private val namespaceInfo = namespaceInfo.associate { it.name to it.strategy }

    fun resolveNamespaceStrategy(node: ModuleDeclaration): NamespaceStrategy? {
        val name = extractNamespaceName(node)
        val detailedName = name.joinToString(separator = ".") { it.detailedName }

        return namespaceInfo[detailedName]
    }
}

class NamespaceInfoPlugin(namespaceInfo: NamespaceInfo) : ConverterPlugin<Node> {
    private val namespaceInfoService = NamespaceInfoService(namespaceInfo)

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun render(node: Node, context: ConverterContext, next: Render<Node>) = null

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()

    override fun setup(context: ConverterContext) {
        context.registerService(namespaceInfoServiceKey, this.namespaceInfoService)
    }
}
